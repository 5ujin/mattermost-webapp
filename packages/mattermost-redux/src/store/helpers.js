// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.

export const offlineConfig = {
    effect: (effect, action) => {
        if (typeof effect !== 'function') {
            throw new Error('Offline Action: effect must be a function.');
        } else if (!action.meta.offline.commit) {
            throw new Error('Offline Action: commit action must be preset.');
        }

        return effect();
    }
};
