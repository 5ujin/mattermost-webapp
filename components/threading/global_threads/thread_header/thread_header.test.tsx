// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {ComponentProps} from 'react';

import {mountWithIntl} from 'tests/helpers/intl-test-helper';

import FollowButton from 'components/threading/common/follow_button';
import ThreadMenu from 'components/threading/global_threads/thread_menu';

import ThreadHeader from './thread_header';

describe('components/threading/global_threads/thread_header', () => {
    let props: ComponentProps<typeof ThreadHeader>;

    beforeEach(() => {
        props = {
            isFollowing: false,
            isSaved: false,
            channelName: 'This is a channel name',
            hasUnreads: false,
            actions: {
                follow: jest.fn(),
                unFollow: jest.fn(),
                openInChannel: jest.fn(),
                markRead: jest.fn(),
                markUnread: jest.fn(),
                save: jest.fn(),
                unSave: jest.fn(),
                copyLink: jest.fn(),
            },
        };
    });

    test('should match snapshot', () => {
        const wrapper = mountWithIntl(
            <ThreadHeader {...props}/>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should support follow', () => {
        const wrapper = mountWithIntl(
            <ThreadHeader {...props}/>,
        );

        wrapper.find(FollowButton).find('button').simulate('click');
        expect(props.actions.follow).toHaveBeenCalled();
    });
    test('should support unfollow', () => {
        props.isFollowing = true;
        const wrapper = mountWithIntl(
            <ThreadHeader {...props}/>,
        );

        wrapper.find(FollowButton).find('button').simulate('click');
        expect(props.actions.unFollow).toHaveBeenCalled();
    });

    test('should support openInChannel', () => {
        props.isFollowing = true;
        const wrapper = mountWithIntl(
            <ThreadHeader {...props}/>,
        );

        wrapper.find('h3 button').simulate('click');
        expect(props.actions.openInChannel).toHaveBeenCalled();
    });

    test('should pass required props to ThreadMenu', () => {
        const wrapper = mountWithIntl(
            <ThreadHeader
                {...props}
            />,
        );
        const {actions} = props;

        // verify ThreadMenu received transient/required props
        new Map<string, any>([
            ['hasUnreads', expect.any(Boolean)],
            ['isFollowing', props.isFollowing],
            ['isSaved', props.isSaved],
            ['actions.follow', actions.follow],
            ['actions.unFollow', actions.unFollow],
            ['actions.openInChannel', actions.openInChannel],
            ['actions.markRead', actions.markRead],
            ['actions.markUnread', actions.markUnread],
            ['actions.save', actions.save],
            ['actions.unSave', actions.unSave],
            ['actions.copyLink', actions.copyLink],
        ]).forEach((val, prop) => {
            expect(wrapper.find(ThreadMenu).props()).toHaveProperty(prop, val);
        });
    });
});

