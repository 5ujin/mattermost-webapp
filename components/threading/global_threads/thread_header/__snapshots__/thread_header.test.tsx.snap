// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/threading/global_threads/thread_header should match snapshot 1`] = `
<Memo(ThreadHeader)
  actions={
    Object {
      "copyLink": [MockFunction],
      "follow": [MockFunction],
      "markRead": [MockFunction],
      "markUnread": [MockFunction],
      "openInChannel": [MockFunction],
      "save": [MockFunction],
      "unFollow": [MockFunction],
      "unSave": [MockFunction],
    }
  }
  channelName="This is a channel name"
  hasUnreads={false}
  isFollowing={false}
  isSaved={false}
>
  <header
    className="ThreadHeader"
  >
    <h3>
      <span
        className="separated"
      >
        Thread
      </span>
      <Memo(Button)
        className="separated"
        onClick={[MockFunction]}
      >
        <button
          className="Button Button___transparent separated"
          onClick={[MockFunction]}
        >
          <span
            className="Button_label"
          >
            This is a channel name
          </span>
        </button>
      </Memo(Button)>
    </h3>
    <div
      className="spacer"
    />
    <Memo(FollowButton)
      follow={[MockFunction]}
      isFollowing={false}
      unFollow={[MockFunction]}
    >
      <Memo(Button)
        isActive={false}
        onClick={[MockFunction]}
      >
        <button
          className="Button Button___transparent"
          onClick={[MockFunction]}
        >
          <span
            className="Button_label"
          >
            Follow
          </span>
        </button>
      </Memo(Button)>
    </Memo(FollowButton)>
    <Memo(ThreadMenu)
      actions={
        Object {
          "copyLink": [MockFunction],
          "follow": [MockFunction],
          "markRead": [MockFunction],
          "markUnread": [MockFunction],
          "openInChannel": [MockFunction],
          "save": [MockFunction],
          "unFollow": [MockFunction],
          "unSave": [MockFunction],
        }
      }
      hasUnreads={false}
      isFollowing={false}
      isSaved={false}
    >
      <MenuWrapper
        animationComponent={[Function]}
        className=""
        stopPropagationOnToggle={true}
      >
        <div
          className="MenuWrapper "
          onClick={[Function]}
        >
          <Memo(SimpleTooltip)
            content="More Actions"
            id="threadActionMenu"
          >
            <OverlayTrigger
              defaultOverlayShown={false}
              delayShow={500}
              overlay={
                <Tooltip
                  bsClass="tooltip"
                  className="hidden-xs"
                  id="threadActionMenu"
                  placement="top"
                >
                  More Actions
                </Tooltip>
              }
              placement="top"
              trigger={
                Array [
                  "hover",
                  "focus",
                ]
              }
            >
              <OverlayTrigger
                defaultOverlayShown={false}
                delayShow={500}
                overlay={
                  <OverlayWrapper
                    bsClass="tooltip"
                    className="hidden-xs"
                    id="threadActionMenu"
                    intl={
                      Object {
                        "defaultFormats": Object {},
                        "defaultLocale": "en",
                        "formatDate": [Function],
                        "formatDateToParts": [Function],
                        "formatDisplayName": [Function],
                        "formatList": [Function],
                        "formatMessage": [Function],
                        "formatNumber": [Function],
                        "formatNumberToParts": [Function],
                        "formatPlural": [Function],
                        "formatRelativeTime": [Function],
                        "formatTime": [Function],
                        "formatTimeToParts": [Function],
                        "formats": Object {},
                        "formatters": Object {
                          "getDateTimeFormat": [Function],
                          "getDisplayNames": [Function],
                          "getListFormat": [Function],
                          "getMessageFormat": [Function],
                          "getNumberFormat": [Function],
                          "getPluralRules": [Function],
                          "getRelativeTimeFormat": [Function],
                        },
                        "locale": "en",
                        "messages": Object {},
                        "onError": [Function],
                        "textComponent": "span",
                        "timeZone": "Etc/UTC",
                        "wrapRichTextChunksInFragment": undefined,
                      }
                    }
                    placement="top"
                  >
                    More Actions
                  </OverlayWrapper>
                }
                placement="top"
                trigger={
                  Array [
                    "hover",
                    "focus",
                  ]
                }
              >
                <Memo(Button)
                  className="Button___icon Button___large"
                  onBlur={[Function]}
                  onClick={null}
                  onFocus={[Function]}
                  onMouseOut={[Function]}
                  onMouseOver={[Function]}
                >
                  <button
                    className="Button Button___transparent Button___icon Button___large"
                    onBlur={[Function]}
                    onClick={null}
                    onFocus={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                  >
                    <span
                      className="Button_label"
                    >
                      <i
                        className="Icon icon-dots-vertical"
                      />
                    </span>
                  </button>
                </Memo(Button)>
              </OverlayTrigger>
            </OverlayTrigger>
          </Memo(SimpleTooltip)>
          <MenuWrapperAnimation
            show={false}
          >
            <CSSTransition
              classNames="MenuWrapperAnimation"
              enter={true}
              exit={true}
              in={false}
              mountOnEnter={true}
              timeout={80}
              unmountOnExit={true}
            >
              <Transition
                appear={false}
                enter={true}
                exit={true}
                in={false}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={80}
                unmountOnExit={true}
              />
            </CSSTransition>
          </MenuWrapperAnimation>
        </div>
      </MenuWrapper>
    </Memo(ThreadMenu)>
  </header>
</Memo(ThreadHeader)>
`;
